[AArch64] Skip over shadow space for ARM64EC entry thunk variadic calls

bylaws opened the pull request:
When in an entry thunk the x64 SP is passed in x4 but this cannot be directly passed through since x64 varargs calls have a 32 byte shadow store at SP followed by the in-stack parameters. ARM64EC varargs calls on the other hand expect x4 to point to the first in-stack parameter.

CC: @efriedma-quic @cjacek @mstorsjo

bylaws commented:
/cherry-pick dff9bb50c75bcc781a6b8414fa33c4c8a0dfd2da

efriedma-quic reviewed (commented):
In general, I attempted to keep the meaning of IR inside the thunks as close as possible to "normal" IR: following the same semantics, just with a slightly exotic calling convention.  So someone who doesn't understand the rules here can just ignore the calling convention stuff, and understand what the IR is doing.

Here, you're violating that rule: you're making a vararg function argument passing work the wrong way.  Outside of musttail calls, vararg lists don't normally get forwarded from the caller to the callee.  The vararg list of a call consists of arguments explicitly passed beyond the end of the fixed argument list.

I see two ways to solve this:
- We try to add some way to model "vararg list forwarding" to LLVM IR.  This doesn't currently exist, so a bit complicated to add.
- We don't model the caller or the callee as "varargs"; instead, we pass x4 and x5 explicitly.  Basically, add "inreg" markings (or something like that) to the arguments representing x4 and x5, then mess with CC_AArch64_Arm64EC_Thunk using CCIfInReg to pass those values in registers instead of memory.

bylaws commented:
> In general, I attempted to keep the meaning of IR inside the thunks as close as possible to "normal" IR: following the same semantics, just with a slightly exotic calling convention. So someone who doesn't understand the rules here can just ignore the calling convention stuff, and understand what the IR is doing.
> 
> Here, you're violating that rule: you're making a vararg function argument passing work the wrong way. Outside of musttail calls, vararg lists don't normally get forwarded from the caller to the callee. The vararg list of a call consists of arguments explicitly passed beyond the end of the fixed argument list.
> 
> I see two ways to solve this:
> 
> * We try to add some way to model "vararg list forwarding" to LLVM IR.  This doesn't currently exist, so a bit complicated to add.
> * We don't model the caller or the callee as "varargs"; instead, we pass x4 and x5 explicitly.  Basically, add "inreg" markings (or something like that) to the arguments representing x4 and x5, then mess with CC_AArch64_Arm64EC_Thunk using CCIfInReg to pass those values in registers instead of memory.

I'm conflicted on this, it would surely be the easiest but I don't think it makes much sense to break the 1-1 mapping between x86 args and entry thunk args, modelling x4 as an argument when it isn't. Using the regular CC to call a varargs CC call also feels quite counter-intuitive. The former seems like a better idea, but modelling both zeroing x5 and offsetting x4 in a generic way seems like it would need something beyond just basic forwarding. Did you have any concrete ideas for handling that?

efriedma-quic commented:
> but I don't think it makes much sense to break the 1-1 mapping between x86 args and entry thunk args, modelling x4 as an argument when it isn't. Using the regular CC to call a varargs CC call also feels quite counter-intuitive.

AArch64Arm64ECCallLowering::getThunkArgTypes already completely throws away the type of varargs functions.  Doing further rewrites doesn't seem like it's making anything worse.

> The former seems like a better idea, but modelling both zeroing x5 and offsetting x4 in a generic way seems like it would need something beyond just basic forwarding. Did you have any concrete ideas for handling that?

I guess you'd make the argument list completely empty (except for sret), mark the call with a "thunk" attribute like we do for musttail varargs forwarding, then check for the "thunk" marking in isel and add the arguments.  I don't really like introducing a whole new kind of varargs forwarding just for this one case, though.

bylaws committed:
[AArch64] Skip over shadow space for ARM64EC entry thunk variadic calls

When in an entry thunk the x64 SP is passed in x4 but this cannot be directly
passed through since x64 varargs calls have a 32 byte shadow store at SP
followed by the in-stack parameters. ARM64EC varargs calls on the other hand
expect x4 to point to the first in-stack parameter.

bylaws commented:
@efriedma-quic how does this look?

efriedma-quic reviewed (approved):
LGTM

efriedma-quic commented:
It looks like the regression tests are failing on the buildbots?

dpaoliello commented:
> It looks like the regression tests are failing on the buildbots?

`arm64ec-entry-thunks.ll` failed, I guess the baseline needs to be updated?

bylaws commented:
@efriedma-quic forgot to push my fix for that after I did it, should be fine now

dpaoliello commented:
@bylaws it's passing on Windows, but still failing when run on Linux - weird...
